service/frontend/workflowHandler.go

```diff
@@ -3502,3 +3538,105 @@ func (wh *WorkflowHandler) checkNamespaceMatch(requestNamespace string, tokenNam
 func (wh *WorkflowHandler) metricsScope(ctx context.Context) metrics.Scope {
 	return interceptor.MetricsScope(ctx, wh.GetLogger())
 }
+
+func (wh *WorkflowHandler) makeFakeContinuedAsNewEvent(
+	ctx context.Context,
+	namespace string,
+	namespaceID string,
+	execution *commonpb.WorkflowExecution,
+	currentBranchToken []byte,
+	lastEvent *historypb.HistoryEvent,
+) (*historypb.HistoryEvent, error) {
+	switch lastEvent.EventType {
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED:
+		if lastEvent.GetWorkflowExecutionCompletedEventAttributes().GetNewExecutionRunId() == "" {
+			return nil, nil
+		}
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_FAILED:
+		if lastEvent.GetWorkflowExecutionFailedEventAttributes().GetNewExecutionRunId() == "" {
+			return nil, nil
+		}
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT:
+		if lastEvent.GetWorkflowExecutionTimedOutEventAttributes().GetNewExecutionRunId() == "" {
+			return nil, nil
+		}
+	default:
+		return nil, nil
+	}
+
+	// We need to replace the last event. Get the run's start event so we can pull some fields out of it.
+	history, _, err := wh.getHistory(
+		wh.metricsScope(ctx),
+		namespaceID,
+		namespace,
+		*execution,
+		common.FirstEventID,
+		common.FirstEventID+1,
+		1,
+		nil,
+		nil,
+		currentBranchToken,
```

Comment:
@wxing1292 can you confirm `currentBranchToken` here can be used for reading the first event?  This is the continuation token received from the client.  Will this work across multiple resets?

Reply:
If this is an issue, we could make the new behavior also conditional on the client not passing a continuation token (in which case we'll be using the branch token we get from mutable state). The queries that we care about don't pass one.

Reply:
it works: https://github.com/temporalio/temporal/blob/v1.12.1/common/persistence/tests/history_store_test.go#L186-L227